builder/qemu/config.go

```diff
@@ -65,6 +65,96 @@ type QemuImgArgs struct {
 	Resize  []string `mapstructure:"resize" required:"false"`
 }
 
+// QemuSMPConfig sets the smp configuration option for the Qemu command-line
+//
+// The smp option sets the number of vCPUs to expose to the VM, the final
+// number of available vCPUs is sockets*cores*threads.
+type QemuSMPConfig struct {
+	// The number of virtual cpus to use when building the VM.
+	//
+	// If undefined, the value will either be `1`, or the product of
+	// `sockets * cpus * threads`
+	//
+	// If both any topology element, or this is set, the final value will be the
+	// largest of the two.
+	//
+	// If the cpu count is the only thing specified, qemu's default behaviour
+	// regarding topology will be applied.
+	// The behaviour depends on the version of qemu; before version 6.2, sockets
+	// were preferred to cores, from version 6.2 onwards, cores are preferred.
+	CpuCount int `mapstructure:"cpus" required:"false"`
+	// The number of sockets to use when building the VM.
+	//  The default is `1` socket.
+	//  The socket count must not be higher than the CPU count.
+	SocketCount int `mapstructure:"sockets" required:"false"`
+	// The number of cores per CPU to use when building the VM.
+	//  The default is `1` core per CPU.
+	CoreCount int `mapstructure:"cores" required:"false"`
+	// The number of threads per core to use when building the VM.
+	//  The default is `1` thread per core.
+	ThreadCount int `mapstructure:"threads" required:"false"`
+}
+
+// getDefaultCmdLine returns the smp command-line argument if defined.
+func (c QemuSMPConfig) getDefaultCmdLine() string {
+	totalVCpus := c.getMaxCPUs()
+	if totalVCpus == 0 && c.CpuCount == 0 {
+		return "1"
+	}
+
+	cpuCount := c.CpuCount
+
+	if cpuCount == 0 {
+		log.Printf("CPU count at default value, setting to topology maximum: %d", totalVCpus)
+		cpuCount = totalVCpus
+	}
+
+	if totalVCpus > cpuCount {
+		log.Print("CPU count lower than what's described in topology." +
+			"This will negatively impact performance.")
+	}
+
+	if cpuCount > totalVCpus && totalVCpus != 0 {
+		log.Printf("CPU count greater than what topology allows, setting to %d", totalVCpus)
+		cpuCount = totalVCpus
+	}
+
+	smpStr := fmt.Sprintf("%d", cpuCount)
+	if c.SocketCount > 0 {
+		smpStr = fmt.Sprintf("%s,sockets=%d", smpStr, c.SocketCount)
+	}
+	if c.CoreCount > 0 {
+		smpStr = fmt.Sprintf("%s,cores=%d", smpStr, c.CoreCount)
+	}
+	if c.ThreadCount > 0 {
+		smpStr = fmt.Sprintf("%s,threads=%d", smpStr, c.ThreadCount)
+	}
+
+	return smpStr
+}
+
+func (c QemuSMPConfig) getMaxCPUs() int {
+	totalVCPUs := c.SocketCount
+
+	if c.CoreCount > 0 {
+		if totalVCPUs == 0 {
+			totalVCPUs = c.CoreCount
+		} else {
+			totalVCPUs *= c.CoreCount
+		}
+	}
+
+	if c.ThreadCount > 0 {
+		if totalVCPUs == 0 {
+			totalVCPUs = c.ThreadCount
+		} else {
+			totalVCPUs *= c.ThreadCount
+		}
+	}
+
+	return totalVCPUs
```

Comment:
nit: The nested if statements is a little confusing. Does something like the following make it clearer? 
```suggestion
	totalVCPUs := c.SocketCount

        if totalVCPUs > 0 && c.CoreCount > 0 {
          totalVCPUs *= c.CoreCount
        } 
        
        // If number of sockets were not provided take the number of cores
        if totalVCPUs == 0 {
            totalVCPUs = c.CoreCount
        }
        
        // CPUCount, SocketCount, and ThreadCount were set return the final product of all.
        if totalVCPUs > 0 && c.ThreadCount > 0 {
               return totalVCPUs *= c.ThreadCount
        } 
        
        // If nothing else was set take the value set for thread count
	if totalVCPUs == 0 {
		return c.ThreadCount
	}

	return totalVCPUs
```